encoder/vaapiencoder_vp9.cpp

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (C) 2014 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "vaapiencoder_vp9.h"
+#include "common/scopedlogger.h"
+#include "common/common_def.h"
+#include "vaapi/vaapicontext.h"
+#include "vaapi/vaapidisplay.h"
+#include "vaapicodedbuffer.h"
+#include "vaapiencpicture.h"
+#include "vaapiencoder_factory.h"
+#include <algorithm>
+
+namespace YamiMediaCodec {
+
+// each frame can have at most 3 reference frames, golden, alt and last
+enum maxSizeValues {
+    kMaxReferenceFrames = 8,
+    kMaxWidth = 8192,
+    kMaxHeight = kMaxWidth
+};
+
+enum VP9QPValues { kMinQPValue = 9, kDefaultQPValue = 60, kMaxQPValue = 127 };
+
+enum frameType { kKeyFrame, kInterFrame };
+
+enum levelValues { kDefaultSharpnessLevel, kDefaultFilterLevel = 10 };
+
+class VaapiEncPictureVP9 : public VaapiEncPicture {
+public:
+    VaapiEncPictureVP9(const ContextPtr& context, const SurfacePtr& surface,
+                       int64_t timeStamp)
+        : VaapiEncPicture(context, surface, timeStamp)
+    {
+        return;
+    }
+
+    VAGenericID getCodedBufferID() { return m_codedBuffer->getID(); }
+};
+
+VaapiEncoderVP9::VaapiEncoderVP9()
+    : m_frameCount(0)
+{
+    m_videoParamCommon.profile = VAProfileVP9Profile0;
+    m_videoParamCommon.rcParams.minQP = kMinQPValue;
+    m_videoParamCommon.rcParams.maxQP = kMaxQPValue;
+    m_videoParamCommon.rcParams.initQP = kDefaultQPValue;
+
+    //add extra surfaces to operate due to kMaxReferenceFrames
+    //vaapi_encoder class will create 5 extra surfaces already
+    m_maxOutputBuffer = kMaxReferenceFrames;
+}
+
+VaapiEncoderVP9::~VaapiEncoderVP9() {}
+
+YamiStatus VaapiEncoderVP9::getMaxOutSize(uint32_t* maxSize)
+{
+    FUNC_ENTER();
+    *maxSize = m_maxCodedbufSize;
+    return YAMI_SUCCESS;
+}
+
+void VaapiEncoderVP9::resetParams()
+{
+    m_maxCodedbufSize = width() * height() * 3 / 2;
+
+    // adding extra padding.  In particular small resolutions require more
+    // space depending on other quantization parameters during execution. The
+    // value below is a good compromise
+    m_maxCodedbufSize += 0x1000;
```

Comment:
@xuguangxin , I noted you changed the VP8_HEADER_MAX_SIZE from 0x1000 to 0x4000 in the commit https://github.com/01org/libyami/commit/07bba09ade99fb4cc4a01677d3ed35a768cfb8b1. Coudld you tell me why?